### PR TITLE
Use generated serialization for NSData

### PR DESCRIPTION
**Pull request by sheeparegreat**
#### 0ee026d9d1d0e2a92dd329244493d8742ccef070
<pre>
Use generated serialization for NSData
<a href="https://bugs.webkit.org/show_bug.cgi?id=263378">https://bugs.webkit.org/show_bug.cgi?id=263378</a>
rdar://117206485

Reviewed by Brady Eidson and Alex Christensen.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::encodeDataInternal):
(IPC::decodeDataInternal):
* Source/WebKit/Shared/Cocoa/CoreIPCData.h: Added.
(WebKit::CoreIPCData::CoreIPCData):
(WebKit::CoreIPCData::createData const):
(WebKit::CoreIPCData::get const):
* Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/269766@main">https://commits.webkit.org/269766@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c31c327fe48ee474fb8ecb76b87d399f6ff706ee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23387 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1501 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24510 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21605 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23940 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1064 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20253 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26151 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/870 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27289 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21330 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25160 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/856 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18596 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/813 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5612 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1268 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1117 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    CoreIPCData(const IPC::DataReference& d)
+        : m_data(adoptCF(CFDataCreate(0, d.data(), d.size())))
+        , m_ref(CFDataGetBytePtr(m_data.get()), CFDataGetLength(m_data.get()))
+    {
+    }
+
+    RetainPtr<CFDataRef> takeData()
+    {
+        return WTFMove(m_data);
+    };
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: unneeded semicolon.

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+public:
+    CoreIPCData();
+
+    CoreIPCData(CFDataRef d)
+        : m_data()
+        , m_ref(CFDataGetBytePtr(d), CFDataGetLength(d))
+    {
+    }
+
+    CoreIPCData(const IPC::DataReference& d)
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: Why is this needed?  Seems suspicious.  It looks like unneeded data copies.

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if PLATFORM(COCOA)
+
+#import "DataReference.h"
+
+#import <Foundation/Foundation.h>
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: This probably shouldn't need Foundation, just CoreFoundation.h

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#import <cstdint>
+#import <span>
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: This is redundant with DataReference.h

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    RetainPtr<CFDataRef> takeData()
+    {
+        return WTFMove(m_data);
+    };
+
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCData, void>;
+    RetainPtr<CFDataRef> m_data;
+    IPC::DataReference m_ref;
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: This actually doesn't need to be stored.  Just make a function like this:
IPC::DataReference asDataReference() { ... }

Then make a constructor that takes a span/DataReference.  Then you only serialize the result of calling this function, and then you don't need to use NotSerialized, which is usually an indication that something needs to be refactored or fixed.

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+#import <Foundation/Foundation.h>
+#import <cstdint>
+#import <span>
+#import <wtf/RetainPtr.h>
+
+namespace WebKit {
+
+class CoreIPCData {
+public:
+    CoreIPCData();
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: Remove.

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    CoreIPCData(CFDataRef d)
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: This constructor needs to set the RetainPtr.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h (hunk 2 above)
Comment (sheeparegreat) [reply Source/WebKit/Shared/Cocoa/CoreIPCData.h]: Because the decoder will make an IPC::DataReference and call into this constructor

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h (hunk 5 above)
Comment (sheeparegreat) [reply Source/WebKit/Shared/Cocoa/CoreIPCData.h]: A single member now which is an IPC::DataReference

**Review by beidson — APPROVED** [review]

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    {
+    }
+
+    RetainPtr<CFDataRef> createData()
+    {
+        return { adoptCF(CFDataCreate(0, m_ref.data(), m_ref.size())) };
+    }
+
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCData, void>;
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: I think it would be even nicer if you made an accessor.  This isn't a great reason to use "friend"

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+namespace WebKit {
+
+class CoreIPCData {
+public:
+    CoreIPCData(const IPC::DataReference& d)
+        : m_ref(d)
+    {
+    }
+
+    RetainPtr<CFDataRef> createData()
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: const

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        return { adoptCF(CFDataCreate(0, m_ref.data(), m_ref.size())) };
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: { and } unnecessary

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCData, void>;
+    IPC::DataReference m_ref;
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: We usually try to avoid abbreviations in our member names.
m_reference
I think it might be even better to just use a std::span (which is interchangeable with DataReference) and call this m_span

File: Source/WebKit/Shared/Cocoa/CoreIPCData.h [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+#import "DataReference.h"
+
+#import <CoreFoundation/CoreFoundation.h>
+#import <wtf/RetainPtr.h>
+
+namespace WebKit {
+
+class CoreIPCData {
+public:
+    CoreIPCData(const IPC::DataReference& d)
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/CoreIPCData.h]: d -> data
More meaningful parameter names.

**Review by achristensen07 — APPROVED** [review]

**Comment by AtariDreams** [comment]
You have to re-add the merge queue thing again @beidson
